### DISCUSSION TITLE
Magic crossfade

### DISCUSSION THREAD

**Discussion by Sarabanga** [savonet/liquidsoap#2169] [Q&A]
Problem -
the air goes in blocks of 1 jingle and 2 songs.
crossfade via meta tags is configured to transition only between songs. There is no crossfade between the jingle and the song.
But after a while, the crossfade starts working with the jingle too. It may also stop working between songs.
What am I doing wrong? 

```
out = output.icecast(
host = "127.0.0.1",
port = 8000,
user = "source",
password = "superpassword",
name = "Best Hits of 60-90",
genre = "Rock,Pop",
url = "http://default.site",
encoding = "UTF-8"
)

tech = "/stream/scripts"

set("log.file.path","#{tech}/liquidsoap.log")
set("log.level", 3)

promo_dir = "/stream/radio/promo"
mus_main_dir = "/stream/radio/music/mainstream"
mus_blues_dir = "/stream/radio/music/blues"
jin_all_dir = "/stream/radio/jingles"


day_music = playlist (mode="randomize", reload_mode="rounds", loop=true, reload = 360, "#{mus_main_dir}")
night_music = playlist (mode="randomize", reload_mode="rounds", loop=true, reload = 360, "#{mus_blues_dir}")
jingles = playlist (mode="random", reload_mode="rounds",reload = 360, "#{jin_all_dir}")
promo = playlist (mode="random", reload_mode="rounds",reload = 360, "#{promo_dir}")
alarm = once(single ("/stream/radio/00.m4a"))



def source_tag(s,tag) =
  def f(_)
    [("source_tag",(tag:string))]
  end
  map_metadata(id=tag,insert_missing=true,f,s)
end

day_music = source_tag(day_music, "music")
night_music = source_tag(night_music, "music")
jingles = source_tag(jingles, "jingles")


def transition(a,b)
  if a.metadata["source_tag"] != "music" or  b.metadata["source_tag"] != "music" then
    sequence([a.source, b.source])
  else
    cross.smart(a, b)
  end
end




def append_title(m) =
  artist = m["artist"]
  title = m["title"]
  year = m["date"]
  [("artist","#{artist}"),("title","#{title} (#{year})")]
end

day_music = map_metadata(append_title, day_music)
night_music = map_metadata(append_title, night_music)



day_efir = rotate (weights = [1, 2], [jingles , day_music])
night_efir = rotate (weights = [1, 2], [jingles, night_music])


radio = switch (track_sensitive = false,
[
  ({23h00m}, alarm),
  ({ true }, day_efir)
])


radio = cross(duration=10., transition, radio)
radio = replaygain(radio)
radio = bass_boost(radio)
radio = normalize(radio)


out(
 %fdkaac(channels=2, samplerate=44100, bandwidth="auto", bitrate=128, afterburner=true, aot="mpeg4_he_aac_v2", transmux="adts", sbr_mode=false),
 description = "AAC Stream",
 mount = "stream",
 mksafe(radio)
)
```

2022/01/24 18:01:34 [decoder:2] Decoding "/stream/radio/jingles/Simply the best - Duke.mp3" ended: Ffmpeg_decoder.End_of_file.
2022/01/24 18:01:34 [mainstream:3] Prepared "/stream/radio/music/mainstream/Scorpions - Wind Of Change.m4a" (RID 5).
2022/01/24 18:01:34 [switch_0:3] Switch to map_metadata_3 with forgetful transition.
2022/01/24 18:01:34 [cross_0:3] Analysis: -79.436455dB / -15.248305dB (9.96s / 9.96s)
2022/01/24 18:01:34 [cross.smart:3] new >= old + margin, old <= medium and new <= high.
2022/01/24 18:01:34 [cross.smart:3] Do not fade if it's already very low.
2022/01/24 18:01:34 [cross.smart:3] Transition: crossed, no fade.
2022/01/24 18:06:40 [decoder:2] Decoding "/stream/radio/music/mainstream/Scorpions - Wind Of Change.m4a" ended: Ffmpeg_decoder.End_of_file.
2022/01/24 18:06:40 [mainstream:3] Prepared "/stream/radio/music/mainstream/Smokie - Living Next Door To Alice.m4a" (RID 7).
2022/01/24 18:06:41 [cross_0:3] Analysis: -91.637780dB / -32.372866dB (9.96s / 9.96s)
2022/01/24 18:06:41 [cross.smart:3] new >= old + margin, old <= medium and new <= high.
2022/01/24 18:06:41 [cross.smart:3] Do not fade if it's already very low.
2022/01/24 18:06:41 [cross.smart:3] Transition: crossed, no fade.
2022/01/24 18:09:59 [decoder:2] Decoding "/stream/radio/music/mainstream/Smokie - Living Next Door To Alice.m4a" ended: Ffmpeg_decoder.End_of_file.
2022/01/24 18:09:59 [jingles:3] Prepared "/stream/radio/jingles/(Robert) Now On Air.mp3" (RID 6).
2022/01/24 18:09:59 [switch_0:3] Switch to jingles.
2022/01/24 18:09:59 [decoder:2] Decoding "/stream/radio/jingles/(Robert) Now On Air.mp3" ended: Ffmpeg_decoder.End_of_file.
2022/01/24 18:09:59 [mainstream:3] Prepared "/stream/radio/music/mainstream/Madonna - Papa Don't Preach.m4a" (RID 1).
2022/01/24 18:09:59 [switch_0:3] Switch to map_metadata_3 with forgetful transition.
2022/01/24 18:10:00 [cross_0:3] Analysis: -93.386240dB / -15.681720dB (9.96s / 9.96s)
2022/01/24 18:10:00 [cross.smart:3] new >= old + margin, old <= medium and new <= high.
2022/01/24 18:10:00 [cross.smart:3] Do not fade if it's already very low.
2022/01/24 18:10:00 [cross.smart:3] Transition: crossed, no fade.

-------------------------------------------------------------------------------------------------------------------

2022/01/24 18:32:38 [mainstream:3] Prepared "/stream/radio/music/mainstream/Mr Mister - Broken Wings.m4a" (RID 2).
2022/01/24 18:32:38 [switch_0:3] Switch to map_metadata_3 with forgetful transition.
2022/01/24 18:32:38 [cross_0:3] Analysis: -83.584675dB / -12.214977dB (9.96s / 9.96s)
2022/01/24 18:32:38 [cross.smart:3] No transition: using default.
2022/01/24 18:36:11 [decoder:2] Decoding "/stream/radio/music/mainstream/Mr Mister - Broken Wings.m4a" ended: Ffmpeg_decoder.End_of_file.
2022/01/24 18:36:11 [mainstream:3] Prepared "/stream/radio/music/mainstream/Europe - The Final Countdown.m4a" (RID 8).
2022/01/24 18:36:12 [cross_0:3] Analysis: -73.695446dB / -32.120943dB (9.96s / 9.96s)
2022/01/24 18:36:12 [cross.smart:3] new >= old + margin, old <= medium and new <= high.
2022/01/24 18:36:12 [cross.smart:3] Do not fade if it's already very low.
2022/01/24 18:36:12 [cross.smart:3] Transition: crossed, no fade.
2022/01/24 18:41:13 [decoder:2] Decoding "/stream/radio/music/mainstream/Europe - The Final Countdown.m4a" ended: Ffmpeg_decoder.End_of_file.
2022/01/24 18:41:13 [jingles:3] Prepared "/stream/radio/jingles/Hit Web Radio-Greg.mp3" (RID 3).
2022/01/24 18:41:13 [switch_0:3] Switch to jingles.
2022/01/24 18:41:13 [decoder:2] Decoding "/stream/radio/jingles/Hit Web Radio-Greg.mp3" ended: Ffmpeg_decoder.End_of_file.
2022/01/24 18:41:13 [mainstream:3] Prepared "/stream/radio/music/mainstream/Coverdale & Page - Take Me For A Little While.m4a" (RID 4).
2022/01/24 18:41:13 [switch_0:3] Switch to map_metadata_3 with forgetful transition.
2022/01/24 18:41:14 [cross_0:3] Analysis: -91.385456dB / -15.275920dB (9.96s / 9.96s)
2022/01/24 18:41:14 [cross.smart:3] new >= old + margin, old <= medium and new <= high.
2022/01/24 18:41:14 [cross.smart:3] Do not fade if it's already very low.
2022/01/24 18:41:14 [cross.smart:3] Transition: crossed, no fade.

**Comment by Russsgithub** [comment]
Hello @Sarabanga ,

It looks to me as if all is working as it should. ```cross.smart``` is a smart crossfade, it decides if / how to crossfade the incoming and outgoing track depending on their volume. 

In your log I see a few of these

```
2022/01/24 18:06:40 [mainstream:3] Prepared "/stream/radio/music/mainstream/Smokie - Living Next Door To Alice.m4a" (RID 7).
2022/01/24 18:06:41 [cross_0:3] Analysis: -91.637780dB / -32.372866dB (9.96s / 9.96s)
2022/01/24 18:06:41 [cross.smart:3] new >= old + margin, old <= medium and new <= high.
2022/01/24 18:06:41 [cross.smart:3] Do not fade if it's already very low.
2022/01/24 18:06:41 [cross.smart:3] Transition: crossed, no fade.
```
```cross``` calculates the loudness of each track (analysis), and ```cross.smart``` decides not to crossfade the tracks as the outgoing tracks loudness is already too low.

You can crossfade every track by removing the smart method i.e cross.smart(a,b) -> cross(a,b). This will still prevent music -> jingle or jingle -> music crossfading but music -> music will alwats crossfade. You may end up with some crossfades you don't want though.

Russ

**Reply by Russsgithub** [reply]
I can't see any jingle crossfades in the log you posted

**Reply by Sarabanga** [reply]
these are two pieces of the log when the crossfade did not work correctly with last song.

**Reply by Russsgithub** [reply]
Yea, I see, sorry, tired eyes I guess.

The only thing that comes to mind is whether the source_tag metadata still exists at the transition .

Have you tried source_tag after map_metadata

I hope you work it out,

**Reply by Sarabanga** [reply]
tried to change. the same.
a crossfade from "Tom's Diner" to "Hot Stuff" and a "Hit Music Radio-Greg" jingle on top of that.
That bug is present, then it is not.
I will go crazy. 

```
2022/01/26 02:13:26 [decoder:2] Decoding "/stream/radio/music/mainstream/Chris Isaak - Wicked Game.m4a" ended: Ffmpeg_decoder.End_of_file.
2022/01/26 02:13:27 [mainstream:3] Prepared "/stream/radio/music/mainstream/DNA feat. Suzanne Vega - Tom's Diner.m4a" (RID 7).
2022/01/26 02:13:27 [cross_0:3] Analysis: -83.051479dB / -28.134963dB (9.96s / 9.96s)
2022/01/26 02:13:27 [cross.smart:3] new >= old + margin, old <= medium and new <= high.
2022/01/26 02:13:27 [cross.smart:3] Do not fade if it's already very low.
2022/01/26 02:13:27 [cross.smart:3] Transition: crossed, no fade.
2022/01/26 02:17:05 [decoder:2] Decoding "/stream/radio/music/mainstream/DNA feat. Suzanne Vega - Tom's Diner.m4a" ended: Ffmpeg_decoder.End_of_file.
2022/01/26 02:17:05 [jingles:3] Prepared "/stream/radio/jingles/Hit Music Radio-Greg.mp3" (RID 6).
2022/01/26 02:17:05 [switch_0:3] Switch to jingles.
2022/01/26 02:17:05 [decoder:2] Decoding "/stream/radio/jingles/Hit Music Radio-Greg.mp3" ended: Ffmpeg_decoder.End_of_file.
2022/01/26 02:17:05 [mainstream:3] Prepared "/stream/radio/music/mainstream/Donna Summer - Hot Stuff.m4a" (RID 1).
2022/01/26 02:17:05 [switch_0:3] Switch to music with forgetful transition.
2022/01/26 02:17:06 [cross_0:3] Analysis: -63.412131dB / -15.332293dB (9.96s / 9.96s)
2022/01/26 02:17:06 [cross.smart:3] new >= old + margin, old <= medium and new <= high.
2022/01/26 02:17:06 [cross.smart:3] Do not fade if it's already very low.
2022/01/26 02:17:06 [cross.smart:3] Transition: crossed, no fade.
```

**Comment by toots** [comment]
Hi,

I have updated the code for `switch` and all its siblings (`rotate`, `fallback` and `random`) to be better at preserving metadata when switching. Any chance you could try with `v2.0.3-preview`?

**Comment by Sarabanga** [comment]
sometimes this happens.
the volume increases right before the crossfade, or perhaps at the beginning of the crossfade. 
at that time log says 
```
2022/02/02 06:41:51 [mainstream:3] Prepared "/stream/radio/music/mainstream/Elton John - Sorry Seems To Be The Hardest Word.m4a" (RID 18).
2022/02/02 06:41:51 [switch_0:3] Switch to music with forgetful transition.
2022/02/02 06:41:51 [cue_cut_0:3] End of track reached before cue-out point.
2022/02/02 06:41:51 [cue_cut_0:3] Cueing in...
2022/02/02 06:41:51 [cross_0:3] Analysis: -infdB / -infdB (1.52s / 1.52s)
2022/02/02 06:42:06 [cue_cut_0:3] Cueing out...
2022/02/02 06:42:06 [mainstream:3] Prepared "/stream/radio/music/mainstream/Black - Wonderful Life.m4a" (RID 20).
```
